Ks network analysis report 1 (#213)
tari-labs/tari-university: src/network-analysis/probabilistic-attack/probability_of_a_byzantine_takeover_of_the_DAN.md

<style>
        div.highlight {
            width: 95%; 
            word-wrap: break-word;
            background: #EA9999;
            font-size: 1.0em;
            padding: 0.5em;
            color: #000000;
            }
        </style>

# Probability of a Byzantine Takeover of the DAN

- [Introduction](#introduction)
- [Tari Digital Assets Network](#tari-digital-assets-network) 
  - [Kademlia](#kademlia)
  - [Node ID](#node-id)
    - [Bootstrapping a Node](#bootstrapping-a-node)
    - [XOR Metric](#xor-metric)
- [Implementation](#implementation)

>   - [Crude Monte Carlo Simulation](#crude-monte-carlo-simulation-1)

Reviewer: ```suggestion
  - [Crude Monte Carlo Simulation](#crude-monte-carlo-simulation)
```

    - [Proving the Law of Large Numbers](#proving-the-law-of-large-numbers)
    - [Individual Probabilities](#individual-probabilities)
    - [Histogram and Visualization of Distribution](#histogram-and-visualization-of-distribution)
  - [Statistical Calculation](#statistical-calculation)
     - [Variation of Total Nodes](#variation-of-total-nodes)
     - [Variation of Byzantine Fault Tolerance Threshold](#variation-of-byzantine-fault-tolerance-threshold)
     - [Variation of Total Number of Nodes with Committee Size 10](#variation-of-total-number-of-nodes-with-committee-size-10)
     - [Variation of Total Number of Nodes with Committee Size 100](#variation-of-total-number-of-nodes-with-committee-size-100)
     - [Variation of Bad Nodes with Committee Size 10 and 100](#variation-of-bad-nodes-with-committee-size-10-and-100)
- [Conclusions and Remarks](#conclusions-and-remarks)
- [References](#references)

> - [Appendices](#appendices)

Reviewer: ```suggestion
- [Appendices](#appendices)
  - [Appendix A: Definitions of Terms](#appendix-a-definitions-of-terms)
```

- [Contributors](#contributors) 

## Introduction 

This research aims to provide answers to questions posed about the workings of the Tari Digital Assets Network (DAN) 
environment: probabilistic attack vector with regard to the total nodes, compromised nodes, committee size and 
Byzantine Fault-tolerance (BFT) threshold. 

This investigation attempts to answer the following question:

> 

Reviewer: ```suggestion
```

- To establish itself on the network without any known contacts, a node needs to contact at least one bootstrap node, 
  requesting an introduction to the network.
- A node ID is generated for the joining node.
- The new node contacts other nodes it is aware of.
- The new node sends a lookup request with its newly generated node ID.
- The contacted nodes return the nodes they know about that are closest.
- The new nodes are added to the routing table, and contacting begins.
- The process continues until the joining node is unable to locate any closer nodes.

This 
*self-lookup* has two effects:

- it allows the node to learn about nodes closer to itself; and
- it populates other nodes' 
  routing tables with the node's ID [[3]].

#### XOR Metric

The Kademlia paper, published in 2002 [[2]], contained the novel idea of using the XOR operator to determine the 
distance and therefore the arrangement of peers within the network. 

Through the XOR metric, a distance is captured. The lookup procedure allows nodes to locate other nodes, 
given a node ID [[3]].

## Implementation

> The following calculations were written to the [Tari Labs Modelling Repository]((https://github.com/tari-labs/modelling)). 

Reviewer: ```suggestion
The following calculations were done using the [Tari Labs Modelling Repository]((https://github.com/tari-labs/modelling)). 
```

##### Normal Distribution

| Statistical Information |                     | Comparison with <br />Theoretical Mean | &nbsp;&nbsp;Difference Calculated |
| ----------------------- | ------------------- | -------------------------------------- | --------------------------------- |
| Intercept               | 0.6482901778221778  | 0.649474335188621                      | 1.18416E-3                        |
| Standard Deviation      | 0.01507612979811762 |                                        |                                   |

#### Histogram and Visualization of Distribution 

The histogram of randomness highlights the distribution of good and bad nodes selected in each experiment, highlighting the random nature of the experiment. 

<br />
<p align="center"><img src="assets/histogram_of_randomness.png" width="700" /></p>
<br />

| Statistical Information |                   |
| ----------------------- | ----------------- |
| Mean                    | 120,000.0         |
| Median                  | 119,991.0         |
| Mode                    | -                 |
| Standard Deviation      | 346.4313595341606 |

### Statistical Calculation

> 

Reviewer: ```suggestion
Literature about [BFT threshold](../../consensus-mechanisms/BFT-consensus-mechanisms-applications/MainReport.md) advises 
the number of good nodes to be at least $\frac{2}{3} \cdot n+1$, where $n$ is the number of nodes. In the calculations 
that follow, BFT threshold of for example $67$% of N, is implemented with rounding up to ensure that at least that 
fraction is obtained. I this sense $67$% of N simulates $\frac{2}{3} \cdot n+1$.

```

##### Variables

  - N (total number of nodes in the network) = $100$
  - m (number of bad actors) = $60$% of N
  - T (BFT threshold) = $50$%, $55$%, $60$%, $67$% of N
  - n (committee size) = ranging from $1$ to $100$ 

<br />

<p align="center"><img src="assets/variation_of_bft_threshold.png" width="700" /></p>
<br />

The above graph was calculated from [variations of BFT](https://github.com/tari-labs/modelling/blob/master/scenarios/T_variations.py) with [hypergeometric distribution](https://github.com/tari-labs/modelling/blob/master/utils/hyper_dist_prob.py). From a plot of committee size versus probability where the number of nodes remains at $100$ with a change in $T$, the BFT 
threshold, ranging from $50$% to $67$%, it can be seen that: When the BFT threshold is $50$% and $55$% the probability is low when the committee size is small; as the committee size increases, the probability increases, and tends to one. The probability is higher for the case where the BFT threshold is $50$% than when the probability is $55$%. 

When the BFT threshold is $60$%, the probability decreases from $0.63$ to approximately $0.59$, where it remains constant. 

> When the BFT threshold is $65$% and $67$%, the probability decreases from $0.38$ and tends to zero. This confirms the BFT threshold of $67$% as per literature. 

Reviewer: ```suggestion
When the BFT threshold is $65$% and $67$%, the probability decreases from $0.38$ and tends to zero. This confirms the 
BFT threshold of $\frac{2}{3} \cdot n+1$ as per literature.
```

#### Variation of Total Number of Nodes with Committee Size 10

##### Variables

- N (total number of nodes in the network) = ranging from $10$ to $350$
- m (number of bad actors) = $60$% of N
- T (BFT threshold) = $67$% of N
- n (committee size) = $10$

> <br />

Reviewer: Please remove all lines containing `<br />`. Only an empty line is required before and after the html image tags.

> <p align="center"><img src="assets/committee_size_10.png" width="700" /></p>

Reviewer: ```suggestion
<p align="center"><img src="assets/committee_size_10.png" width="900" /></p>
```

<br />

The above graph was calculated from [variations of N with n fixed](https://github.com/tari-labs/modelling/blob/master/scenarios/variation_of_N_n_fixed.py) with [hypergeometric distribution](https://github.com/tari-labs/modelling/blob/master/utils/hyper_dist_prob.py). For the graph showing varying probabilities with respect to the total number of network nodes, where the committee size is $10$, the 
probability dramatically increases when the total nodes is three times more than the committee size and onwards. The probability plateaus at $0.35$. 

#### Variation of Total Number of Nodes with Committee Size 100

##### Variables

- N (total number of nodes in the network) = ranging from $100$ to $1300$
- m (number of bad actors) = $60$% of N
- T (BFT threshold) = $67$% of N
- n (committee size) = $100$

<br />

> <p align="center"><img src="assets/committee_size_100.png" width="700" /></p>

Reviewer: ```suggestion
<p align="center"><img src="assets/committee_size_100.png" width="900" /></p>
```

<br />

The above graph was calculated from [variations of N with n fixed](https://github.com/tari-labs/modelling/blob/master/scenarios/variation_of_N_n_fixed.xlsx) with [hypergeometric distribution](https://github.com/tari-labs/modelling/blob/master/utils/hyper_dist_prob.py). From this and the previous graph, it can be seen that probabilities are significantly lower when the committee size is $100$ compared to $10$. There is an increase in probability 
up to a network size of $700$, albeit, not as steep as the change when the committee size is $10$.  The probability plateaus at $0.08$.

The larger the committee size, the less dramatic changes there are in the probability. 

#### Variation of Bad Nodes with Committee Size 10 and 100

##### Variables

- N (total number of nodes in the network) = ranging from $10$ and $100$ to $50,000$
- m (number of bad actors) = $10$%, $20$%, $30$%, $40$%, $50$%, $60$%, $70$%, $80$% and $90$% of N
- T (BFT threshold) = $67$% of N
- n (committee size) = $10$ and $100$

<br />

> <p align="center"><img src="assets/probability_when_committee_100_20.png" width="600" /></p>

Reviewer: ```suggestion
<p align="center"><img src="assets/probability_when_committee_100_20.png" width="900" /></p>
```

<br />

> <p align="center"><img src="assets/probability_when_committee_100_40.png" width="600" /></p>

Reviewer: ```suggestion
<p align="center"><img src="assets/probability_when_committee_100_40.png" width="900" /></p>
```

<br />

> <p align="center"><img src="assets/probability_when_committee_100_60.png" width="600" /></p>

Reviewer: ```suggestion
<p align="center"><img src="assets/probability_when_committee_100_60.png" width="900" /></p>
```

<br />

> <p align="center"><img src="assets/probability_when_committee_100_90.png" width="600" /></p>

Reviewer: ```suggestion
<p align="center"><img src="assets/probability_when_committee_100_90.png" width="900" /></p>
```

<br />

The above graphs were calculated from [bad node variation where n is 100](https://github.com/tari-labs/modelling/blob/master/scenarios/bad_node_variation_10_10.xlsx.zip) with [hypergeometric distribution](https://github.com/tari-labs/modelling/blob/master/utils/hyper_dist_prob.py). These graphs show varying probabilities when the percentage of bad nodes is $20$, $40$, $60$ and $90$. The value when the probability plateaus is used to construct the graph below for both committee sizes $10$ and $100$. 

<br />

> <p align="center"><img src="assets/probability_bad_nodes_10_100.png" width="700" /></p>

Reviewer: ```suggestion
<p align="center"><img src="assets/probability_bad_nodes_10_100.png" width="850" /></p>
```

## Conclusions and Remarks 

With regards to the Crude Monte Carlo Simulation, at this building block stage, probabilities were calculated and distributions of nodes within the network illustrated.

With regards to the statisical calculation, comments can be made for each of the varied parameters. 

- Total nodes in the network: the smaller the pool of total nodes in the network, the lower the probability of bad actors controlling the network; however, the probability difference is near negligible if the committee size is large. Also, this parameter will be difficult to control, and the network will be ever-increasing. This can be seen from the graph in [Variations of Total Nodes](#variation-of-total-nodes). 

> - BFT threshold: this threshold should be higher than $67$% as per literature. This can be seen from the graph in [Variation of Byzantine Fault-Tolerance Threshold](#variation-of-byzantine-fault-tolerance-threshold)

Reviewer: ```suggestion
- BFT threshold: this threshold should be at least  $\frac{2}{3} \cdot n+1$ as per literature. This can be seen from the graph in 
[Variation of Byzantine Fault-Tolerance Threshold](#variation-of-byzantine-fault-tolerance-threshold)
```

> - Committee size: the larger the committee size, the lower the probability bad actors controlling the network. This can be seen from the graph in [Variation of Total Number of Nodes with Committee Size 10](#variation-of-total-number-of-nodes-with-committee-size-10) and [Variation of Total Number of Nodes with Committee Size 100](#variation-of-total-number-of-nodes-with-committee-size-100).

Reviewer: Please ensure line widths of ~120 are adhered to for easier review comments and comparison of changes - for report in general.
_**Note:** This excludes links, tables and equations, which should not have newline characters in between._
